src/System.Management.Automation/engine/InternalCommands.cs

```diff
@@ -275,164 +495,204 @@ protected override void BeginProcessing()
             {
-                case "ScriptBlockSet":
-                    for (int i = _start; i < _end; i++)
+                if (_arguments != null && _arguments.Length > 0)
+                {
+                    WriteError(GenerateNameParameterError("InputObject", ParserStrings.InvokeMethodOnNull,
+                                                          "InvokeMethodOnNull", _inputObject));
+                }
+                else
+                {
+                    // should process
+                    string propertyAction = string.Format(CultureInfo.InvariantCulture,
+                        InternalCommandStrings.ForEachObjectPropertyAction, _propertyOrMethodName);
+
+                    if (ShouldProcess(_targetString, propertyAction))
                     {
-                        // Only execute scripts that aren't null. This isn't treated as an error
-                        // because it allows you to parameterize a command - for example you might allow
-                        // for actions before and after the main processing script. They could be null
-                        // by default and therefore ignored then filled in later...
-                        if (_scripts[i] != null)
+                        if (Context.IsStrictVersion(2))
+                        {
+                            WriteError(GenerateNameParameterError("InputObject", InternalCommandStrings.InputObjectIsNull,
+                                                                  "InputObjectIsNull", _inputObject));
+                        }
+                        else
                         {
-                            _scripts[i].InvokeUsingCmdlet(
-                                contextCmdlet: this,
-                                useLocalScope: false,
-                                errorHandlingBehavior: ScriptBlock.ErrorHandlingBehavior.WriteToCurrentErrorPipe,
-                                dollarUnder: InputObject,
-                                input: new object[] { InputObject },
-                                scriptThis: AutomationNull.Value,
-                                args: Array.Empty<object>());
+                            // we write null out because:
+                            // PS C:\> $null | ForEach-object {$_.aa} | ForEach-Object {$_ + 3}
+                            // 3
+                            // so we also want
+                            // PS C:\> $null | ForEach-object aa | ForEach-Object {$_ + 3}
+                            // 3
+                            // But if we don't write anything to the pipeline when _inputObject is null,
+                            // the result 3 will not be generated.
+                            WriteObject(null);
                         }
                     }
+                }
 
-                    break;
-                case "PropertyAndMethodSet":
+                return;
+            }
+
+            ErrorRecord errorRecord = null;
```

Comment:
Will this error record be used in all paths below it? My feeling is that an error record variable should only exist on paths where an error must occur. In this case I can't actually find the immediate use of this variable

Reply:
Old code, out of scope for this PR.